### ISSUE TITLE
`docker build` fails to authenticate with private repository

### ISSUE THREAD

**Issue by alexhendel** [docker/buildx#1613]

**Comment by thaJeztah** [comment]
Done!

**Comment by tonistiigi** [comment]
Are you using a remote server and is it possible the registry does additional authentication based on IP? If yes, does setting `BUILDKIT_NO_CLIENT_TOKEN=true` make any difference?

> "***": {

Is this a clean domain like in the reference you are pulling, or URL with some protocol/path?

**Comment by alexhendel** [comment]
> Are you using a remote server and is it possible the registry does additional authentication based on IP? If yes, does setting `BUILDKIT_NO_CLIENT_TOKEN=true` make any difference?
> 

We are using [Sonatype Nexus](https://www.sonatype.com/products/nexus-repository) to host a private registry. I am not sure about the IP-based authentication.

Where would I configure `BUILDKIT_NO_CLIENT_TOKEN=true`?

> > "***": {
> 
> Is this a clean domain like in the reference you are pulling, or URL with some protocol/path?

This value is a clean domain. No path or protocol.

**Comment by alexhendel** [comment]
We did some additional testing which might give some more details. We did try podman and ran:

```bash
podman pull
podman build .
```

This behaves like the previous Docker version, build and pull work fine. Checking the network traffic the only difference we could find was an additional HEAD request with v23 (which was not sent previously and is also not sent by podman).

**Comment by posti-vpk** [comment]
We have experienced the exactly same issue with docker `23.0.1` against private images in GitHubs **_ghcr.io_**-registry. Adding `BUILDKIT_NO_CLIENT_TOKEN=true` fixed the issue for us. Also downgrading to `20.x.x` solved the issue.

**Comment by tgquan67** [comment]
I can confirm that `BUILDKIT_NO_CLIENT_TOKEN=1` works for us (we are using docker 23, and the private registry is artifactory with client certificate required for authentication).

**Comment by krische** [comment]
> I can confirm that `BUILDKIT_NO_CLIENT_TOKEN=1` works for us (we are using docker 23, and the private registry is artifactory with client certificate required for authentication).

@tgquan67 What version of Artifactory are you using? I can't seem to get it to work with Artifactory `7.49.3`.

**Comment by tonistiigi** [comment]
`BUILDKIT_NO_CLIENT_TOKEN=1` suggests that the difference is in some network layer and only the daemon can connect to the auth/registry server. The code itself should be identical.

**Comment by krische** [comment]
So `BUILDKIT_NO_CLIENT_TOKEN=1` is supposed to be an environment variable for the daemon process? Not the client?

**Comment by tonistiigi** [comment]
No, on the client. `BUILDKIT_NO_CLIENT_TOKEN=1 docker buildx build ...`

**Comment by krische** [comment]
> No, on the client. `BUILDKIT_NO_CLIENT_TOKEN=1 docker buildx build ...`

Yup, that's what I was trying but still no luck. What's odd for me is it does seem like my layers are pushed to my Artifactory successfully, but when it tries to push the manifest then Artifactory responds with a `403`. And since I'm on a nearly 6 month old version of Artifactory, I'm not 100% certain the issue with is with buildkit/buidx.

I'll have to update my Artifactory and try again.

**Comment by thaJeztah** [comment]
@tonistiigi related to these?

- https://github.com/containerd/containerd/issues/4982
- https://github.com/containerd/containerd/pull/3913
- https://github.com/containerd/containerd/pull/3912
- https://github.com/containerd/containerd/pull/4983

Last one has further details (it was rejected at the time); https://github.com/containerd/containerd/pull/4983#issuecomment-798727645

**Comment by tgquan67** [comment]
> @tgquan67 What version of Artifactory are you using? I can't seem to get it to work with Artifactory `7.49.3`.

I'm on artifactory 7.41 and it worked. Btw, in my case I only test if it can pull the base image from artifactory, I did not try to push it with buildx.
Maybe disable buildkit with `DOCKER_BUILDKIT=0` and see if it works first, and then check how you authenticate with artifactory.

**Comment by tonistiigi** [comment]
No, these seem to be about HTTP request methods,

**Comment by krische** [comment]
I updated my Artifactory to their newest version, but still no luck. However, I did try `docker login` with the admin account of my Artifactory and that worked. So I don't think my issue is with buildx but something with my Artifactory.

**Comment by krische** [comment]
Yup, it was just a permission issue on my Artifactory that caused the `403` when buildkit was trying to `HEAD` to see if the manifest was already there before the `PUT` to upload it. I fixed the permission issue and it worked. It even works for me without setting `BUILDKIT_NO_CLIENT_TOKEN=1`.

**Comment by tgquan67** [comment]
Now I test it again on docker 24, it works even without `BUILDKIT_NO_CLIENT_TOKEN=1` on the default builder; but if i create a new builder (for example with `docker buildx create --name test --driver=docker-container`), I have no way to make it work when this new builder is in use (with `docker buildx use test`).

**Comment by typoworx-de** [comment]
Same issue here!

**Comment by fredericoschardong** [comment]
The issue persists on docker 24.0.7 and buildx v0.11.2.

**Comment by ryan-ph** [comment]
I'm still seeing this on docker 25.0.3 and buildx v0.12.1

```sh
$ docker --version
Docker version 25.0.3, build 4debf41

$ docker buildx version
github.com/docker/buildx v0.12.1-desktop.4 6996841df2f61988c2794d84d33205368f96c317
```

**Comment by marco-bertelli** [comment]
same issue here

**Comment by whitfin** [comment]
@alexhendel just curious if you ever resolved this on your side? We're using Nexus and facing the same thing.

**Comment by tgquan67** [comment]
Another update: at some point after docker 25 update, it started working in my case (artifactory 7.41) with `BUILDKIT_NO_CLIENT_TOKEN=1`. Also since I have client certificate requirement for my registry, I also have this file at `/etc/buildkitd.toml`:
```
debug = true
[registry."artifactory.myregistry.net"]
  ca=["/opt/myregistry/certs/ca.crt"]
  [[registry."artifactory.myregistry.net".keypair]]
    key="/opt/myregistry/certs/client.key"
    cert="/opt/myregistry/certs/client.pem"
``` 
After this in dockerfiles I can pull from `artifactory.myregistry.net` using credentials I logged in with `docker login artifactory.myregistry.net`.

**Comment by alexhendel** [comment]
> @alexhendel just curious if you ever resolved this on your side? We're using Nexus and facing the same thing. Works with legacy builder, just not BuildKit.

No not directly.

There was an nginx proxy routing our traffic. We did change the nginx config to allow HEAD requests. To us it seems that at some point these HEAD requests where implemented and did not go through with our setup. It did work for our case since.

**Comment by marco-bertelli** [comment]
In my case guys was a GitHub package permission issue, for fix that go into your package detail ---> package settings ---> and check the package permissions, not sure that this is releated to your problem but maybe can help, after that on my side no more auth errors, the strange thing was the error because was not simila to 401 or 403.

**Comment by zerowebcorp** [comment]
Same issue with latest docker.

root@b:~# docker --version
Docker version 27.3.1, build ce12230
root@b:~#

**Comment by thaJeztah** [comment]
@zerowebcorp are you running the `docker` CLI inside a container? Credentials are stored on the client-side (and/or obtained through credential-helpers on the client-side); if you're running from inside a container, make sure you're authenticated inside the container, and that authentication works there (e.g. in a situation where you bind-mount the home-directory or `~/.docker/` directory into a container, the CLI configuration file may contain configuration to use a credential-helper that's present on the host, but not available inside the container.

**Comment by zerowebcorp** [comment]
No. Outside the container. Docker registry hosted remote behind cloudflare. It started working finally when I disabled cloudflare waf.

**Comment by 100tomer** [comment]
> [@zerowebcorp](https://github.com/zerowebcorp) are you running the `docker` CLI inside a container? Credentials are stored on the client-side (and/or obtained through credential-helpers on the client-side); if you're running from inside a container, make sure you're authenticated inside the container, and that authentication works there (e.g. in a situation where you bind-mount the home-directory or `~/.docker/` directory into a container, the CLI configuration file may contain configuration to use a credential-helper that's present on the host, but not available inside the container.

Hi, when I run this locally, it works:
```
docker login...
docker buildx build...
```

However, when I try the same commands inside a Docker container during CI, it fails due to authentication issues. I believe the authentication is fine since a normal Docker push to the same repository works without any problems.

**Comment by ChGen** [comment]
Still relevant for private & corp. registries. I've tried several available tricks from this issue and related but they don't work for push operations. So, for docker push to private insecure registry works only when using default builder with containerd off. If containerd off, but extra builder is created it will be impossible to push and sometimes even to pull from that repo.

**Comment by crazy-max** [comment]
Thanks everyone for the reports and details.

I'm going to close this issue to reset the investigation. This thread now mixes several different failure modes, including registry/proxy setups rejecting `HEAD` requests, WAF behavior, missing or different Docker config in CI/containerized clients, BuildKit registry TLS or insecure-registry config, and registry permission issues. Some of those were resolved by configuration changes, while others were reported against older Docker, Buildx, and BuildKit versions.

Closing this does not mean every private-registry auth failure is considered fixed. It means this issue is no longer actionable as a single tracker.

If you can still reproduce this with current Docker, Buildx, and BuildKit versions, please open a new issue with a fresh minimal repro and include `docker version`, `docker buildx version`, `docker buildx ls`, the exact build command, the builder driver, sanitized registry/proxy details, sanitized Docker config shape, and the full error output.
